dora/tests/src/test/java/alluxio/client/rest/S3ObjectTest.java

```diff
@@ -0,0 +1,412 @@
+  }
+
+  /**
+   * Puts and heads directories.
+   */
+  @Test
+  public void putAndHeadDirectories() throws Exception {
+    final String bucket = "bucket" + AlluxioURI.SEPARATOR;
+    final String objectKey1 = bucket + "folder0" + AlluxioURI.SEPARATOR;
+    final String objectKey2 = objectKey1 + "folder1" + AlluxioURI.SEPARATOR;
+    createBucketTestCase(bucket).checkResponseCode(Status.OK.getStatusCode());
+    createObjectTestCase(objectKey2, EMPTY_OBJECT).checkResponseCode(Status.OK.getStatusCode());
+    headTestCase(objectKey1).checkResponseCode(Status.OK.getStatusCode());
+    headTestCase(objectKey2).checkResponseCode(Status.OK.getStatusCode());
+  }
+
+  /**
+   * Overwrites an existent object.
+   */
+  @Test
+  public void overwriteObject() throws Exception {
+    final String bucket = "bucket";
+    final String objectKey = "bucket/object";
+    final byte[] object = "Hello World!".getBytes();
+    final byte[] object2 = CommonUtils.randomAlphaNumString(Constants.KB).getBytes();
+    createBucketTestCase(bucket).checkResponseCode(Status.OK.getStatusCode());
+    createObjectTestCase(objectKey, object).checkResponseCode(Status.OK.getStatusCode());
+    getTestCase(objectKey).checkResponseCode(Status.OK.getStatusCode()).checkResponse(object);
+    //   This object will be overwritten.
+    createObjectTestCase(objectKey, object2).checkResponseCode(Status.OK.getStatusCode());
+    getTestCase(objectKey).checkResponseCode(Status.OK.getStatusCode()).checkResponse(object2);
+  }
+
+  /**
+   * Puts a directory to a non-existent bucket.
+   */
+  @Test
+  public void putDirectoryToNonExistentBucket() throws Exception {
+    final String bucket = "non-existent-bucket";
+    final String objectKey = "non-existent-bucket/folder/";
+    final byte[] object = new byte[] {};
+    headTestCase(bucket).checkResponseCode(Status.NOT_FOUND.getStatusCode());
+    createObjectTestCase(objectKey, object).checkResponseCode(Status.NOT_FOUND.getStatusCode())
+        .checkErrorCode(S3ErrorCode.Name.NO_SUCH_BUCKET);
+  }
+
+  /**
+   * Puts an object to a non-existent bucket.
+   */
+  @Test
+  public void putObjectToNonExistentBucket() throws Exception {
+    final String bucket = "non-existent-bucket";
+    final String objectKey = "non-existent-bucket/object";
+    final byte[] object = "Hello World!".getBytes();
+
+    headTestCase(bucket).checkResponseCode(Status.NOT_FOUND.getStatusCode());
+    createObjectTestCase(objectKey, object).checkResponseCode(Status.NOT_FOUND.getStatusCode())
+        .checkErrorCode(S3ErrorCode.Name.NO_SUCH_BUCKET);
+  }
+
+  /**
+   * Puts an object with wrong MD5.
+   */
+  @Test
+  public void putObjectWithWrongMD5() throws Exception {
+    final String bucket = "bucket";
+    final String objectKey = "bucket/object";
+    final byte[] object = "Hello World!".getBytes();
+
+    createBucketTestCase(bucket).checkResponseCode(Status.OK.getStatusCode());
+    newTestCase(objectKey, NO_PARAMS,
```

Comment:
how about adding another createObjectTestCase method with the MD5 parameter on top of the existing one? so you can decide to use the self-defined MD5 or the computed MD5.